materialize-firebolt/transactor.go

```diff
@@ -0,0 +1,287 @@
+			Bucket:  t.bucket,
+			Key:     fmt.Sprintf("%s%s.json", t.prefix, randomDocumentKey()),
+			Binding: i,
+		})
+	}
+
+	jsn, err := json.Marshal(checkpoint)
+	if err != nil {
+		return pf.DriverCheckpoint{}, fmt.Errorf("creating checkpoint json: %w", err)
+	}
+	t.checkpoint = checkpoint
+
+	return pf.DriverCheckpoint{DriverCheckpointJson: jsn}, nil
+}
+
+func (t *transactor) projectDocument(spec *pf.MaterializationSpec_Binding, keys tuple.Tuple, values tuple.Tuple) ([]byte, error) {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		var propName = spec.FieldSelection.Keys[i]
+		document[propName] = value
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		var propName = spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+
+	jsonDoc, err := json.Marshal(document)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize the addition document: %w", err)
+	}
+
+	return jsonDoc, nil
+}
+
+// write file to S3 and keep a record of them to be copied to the table
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	awsConfig := aws.Config{
+		Credentials: credentials.NewStaticCredentials(t.awsKeyId, t.awsSecretKey, ""),
+		Region:      &t.awsRegion,
+	}
+	sess := session.Must(session.NewSession(&awsConfig))
+	uploader := s3manager.NewUploader(sess)
+
+	uploads := []s3manager.BatchUploadObject{}
+
+	files := make(map[int][]byte)
+
+	for it.Next() {
+		doc, err := t.projectDocument(t.bindings[it.Binding].spec, it.Key, it.Values)
+		if err != nil {
+			return fmt.Errorf("projecting new store json: %w", err)
+		}
+
+		if _, ok := files[it.Binding]; ok {
+			files[it.Binding] = append(files[it.Binding], byte('\n'))
+			files[it.Binding] = append(files[it.Binding], doc...)
+		} else {
+			files[it.Binding] = doc
+		}
+	}
+
+	for bindingIndex, data := range files {
+		cp := t.checkpoint.Files[bindingIndex]
+
+		uploads = append(uploads, s3manager.BatchUploadObject{
+			Object: &s3manager.UploadInput{
+				Bucket: aws.String(cp.Bucket),
+				Key:    aws.String(cp.Key),
+				Body:   bytes.NewReader(data),
+			},
+		})
+	}
+
+	uploadIterator := s3manager.UploadObjectsIterator{
+		Objects: uploads,
+	}
+
+	err := uploader.UploadWithIterator(aws.BackgroundContext(), &uploadIterator)
+	if err != nil {
+		return fmt.Errorf("uploading files to s3: %w", err)
+	}
+
+	return nil
+}
+
+func (t *transactor) Commit(ctx context.Context) error {
+	return nil
+}
+
+// load stored file into table
+func (t *transactor) Acknowledge(context.Context) error {
+
+	if len(t.checkpoint.Files) < 1 {
+		return nil
+	}
+
+	sourceFileNames := map[int][]string{}
+	for _, item := range t.checkpoint.Files {
+		binding := item.Binding
+		fileName := item.Key
+
+		if _, ok := sourceFileNames[binding]; ok {
```

Comment:
How does this case run? It's not possible for there to be more than one file per binding, correct?

Reply:
Ah yes, my initial implementation saved multiple files for each binding, I forgot to update this 👍🏽